### PR TITLE
feat: flashbots style auth

### PR DESCRIPTION
**Pull request by FrankieIsLost**
Opening up a draft PR to discuss adding[ flashbots-style auth](https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#authentication) to the rpc client. This implementation is based on a tower middleware responsible for signing request bodies and adding the signature to a header. I've also added an example showing how to set up the rpc-client with signing middleware, and sending a request. 

A few questions: 

- should the http client with the signing middleware be abstracted into a separate "authed-client" implementation, or should we expect user to manually set up the signing middleware (like it's done in the example) whenever they want to use the client? 
- where should the layer implementation live? I've added it to a `auth.rs` file in the rpc-api crate, but not sure if this is right. 
- should the auth layer be an optional dependency behind a feature?

### CODE REVIEW

File: crates/mev-share-rpc-api/src/auth.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+impl<S, I> Service<Request<Body>> for FlashbotsSigner<S, I>
+where
+    I: Service<Request<Body>> + Clone + Send + 'static,
+    I::Future: Send,
+    S: Signer + Clone + Send + 'static,
+{
+    type Response = I::Response;
+    type Error = I::Error;
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/auth.rs:58]: we need a bit more error handling here to get rid of the unwraps

>  I::Error: Into<Box<dyn Error + Send + Sync>> + 'static,

and

> type Error = Box<dyn Error + Send + Sync>;

is common, so the service returns a boxed error and the wrapped service's error can be converted into this

File: crates/mev-share-rpc-api/src/auth.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        let clone = self.inner.clone();
+        // wait for service to be ready
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+        let signer = self.signer.clone();
+
+        let (mut parts, body) = request.into_parts();
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/auth.rs:71]: should this restrict what kind of requests are supported, like only POST or only json content type?

File: crates/mev-share-rpc-api/src/auth.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        Box::pin(async move {
+            let body_bytes = hyper::body::to_bytes(body).await.unwrap();
+
+            // sign request body and insert header
+            let signature = signer
+                .sign_message(format!("0x{:x}", H256::from(keccak256(body_bytes.clone()))))
+                .await
+                .unwrap();
+
+            let header_name = HeaderName::from_static("x-flashbots-signature");
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/auth.rs:82]: worth making this a constant, should be possible I think

Comment (mattsse) [inline crates/mev-share-rpc-api/src/auth.rs:82]: should this request fail if there's already a signature? or do nothing?

File: crates/mev-share-rpc-api/src/auth.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            inner,
+        }
+    }
+}
+
+/// Middleware that adds a request header with a signed payload.
+/// For more info, see https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#authentication
+#[derive(Clone)]
+pub struct FlashbotsSigner<S, I> {
+    signer: Arc<S>,
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/auth.rs:47]: I think we can remove the Arc here and instead require that the Signer is clone:

> S: Signer + Clone + Send + 'static,

**Review by mattsse — CHANGES_REQUESTED** [review]
> should the http client with the signing middleware be abstracted into a separate "authed-client" implementation, or

I think we can put this behind the client feature

File: crates/mev-share-rpc-api/src/auth.rs (hunk 1 above)
Comment (FrankieIsLost) [reply crates/mev-share-rpc-api/src/auth.rs]: fixed 👍

File: crates/mev-share-rpc-api/src/auth.rs (hunk 2 above)
Comment (FrankieIsLost) [reply crates/mev-share-rpc-api/src/auth.rs:69]: changed to pass through requests that are not POST or json content type, but let me know if you think we should error out instead

File: crates/mev-share-rpc-api/src/auth.rs (hunk 3 above)
Comment (FrankieIsLost) [reply crates/mev-share-rpc-api/src/auth.rs:97]: changed to just pass through requests that already have a signature

File: crates/mev-share-rpc-api/src/auth.rs (hunk 4 above)
Comment (FrankieIsLost) [reply crates/mev-share-rpc-api/src/auth.rs]: fixed 👍

File: crates/mev-share-rpc-api/src/auth.rs (hunk 3 above)
Comment (FrankieIsLost) [reply crates/mev-share-rpc-api/src/auth.rs]: done

File: crates/mev-share-rpc-api/src/auth.rs [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+impl<S: Clone, I> Layer<I> for FlashbotsSignerLayer<S> {
+    type Service = FlashbotsSigner<S, I>;
+
+    fn layer(&self, inner: I) -> Self::Service {
+        FlashbotsSigner { signer: self.signer.clone(), inner }
+    }
+}
+
+/// Middleware that adds a request header with a signed payload.
+/// For more info, see https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#authentication
```
Comment (mattsse) [inline crates/mev-share-rpc-api/src/auth.rs:43]: this should mention that this will sign the request body and set the flashbots header

**Review by mattsse — CHANGES_REQUESTED** [review]
only a few nits

**Review by mattsse — APPROVED** [review]
lgtm

**Review by mattsse — APPROVED** [review]

**Pull request merged**